### ISSUE TITLE
panic: tail: tail send failed

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#5660]
Run the following go program and materialize will panic in just a few seconds. The program runs a random command (or closes the connection) on a random connection.

```
package main

import (
	"database/sql"
	"fmt"
	"log"
	"math/rand"

	"github.com/lib/pq"
)

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

func run() error {
	rng := rand.New(rand.NewSource(0))
	conns := make(map[int]*sql.DB)

	for {
		connN := rng.Intn(3)
		if _, ok := conns[connN]; !ok {
			connector, err := pq.NewConnector("user=materialize dbname=materialize sslmode=disable port=6875")
			if err != nil {
				return err
			}
			db := sql.OpenDB(connector)
			db.Exec("CREATE TABLE IF NOT EXISTS t (i INT)")
			conns[connN] = db
		}
		db := conns[connN]

		switch rng.Intn(5) {
		case 0:
			fmt.Println("BEGIN")
			if _, err := db.Exec("BEGIN"); err != nil {
				fmt.Println("ERR", err)
			}
		case 1:
			fmt.Println("COMMIT")
			if _, err := db.Exec("COMMIT"); err != nil {
				fmt.Println("ERR", err)
			}
		case 2:
			fmt.Println("INSERT")
			if _, err := db.Exec("INSERT INTO t VALUES ($1), ($2), ($3)", rng.Intn(100), rng.Intn(100), rng.Intn(100)); err != nil {
				fmt.Println("ERR", err)
			}
		case 3:
			fmt.Println("DECLARE")
			if _, err := db.Exec(fmt.Sprintf("DECLARE c%d CURSOR FOR TAIL t", rng.Intn(3))); err != nil {
				fmt.Println("ERR", err)
			}
		case 4:
			fmt.Println("FETCH")
			rows, err := db.Query(fmt.Sprintf("FETCH ALL c%d WITH (timeout = '0s')", rng.Intn(3)))
			if err != nil {
				fmt.Println("ERR", err)
				continue
			}
			rows.Close()
		case 5:
			fmt.Println("CLOSE")
			db.Close()
			delete(conns, connN)
		}
	}
}
```

with:

```
[2m2021-02-08T05:19:57.851506Z[0m [31mERROR[0m panic: timely:work-1: tail send failed: comm error: bincode/io error: io error: Broken pipe (os error 32)
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:569:65
   1: core::ops::function::Fn::call
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:581:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:484:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:153:18
   5: rust_begin_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:483:5
   6: core::panicking::panic_fmt
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:85:14
   7: core::option::expect_none_failed
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/option.rs:1234:5
   8: core::result::Result<T,E>::expect
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:933:23
   9: dataflow::sink::tail::tail::{{closure}}::{{closure}}
             at src/dataflow/src/sink/tail.rs:91:13
  10: timely::dataflow::operators::generic::handles::InputHandle<T,D,P>::for_each
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/handles.rs:92:13
  11: timely::dataflow::operators::generic::handles::FrontieredInputHandle<T,D,P>::for_each
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/handles.rs:136:9
  12: dataflow::sink::tail::tail::{{closure}}
             at src/dataflow/src/sink/tail.rs:37:9
  13: <timely::dataflow::stream::Stream<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::sink::{{closure}}::{{closure}}
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/operator.rs:512:17
  14: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
  15: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/builder_rc.rs:163:26
  16: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/builder_raw.rs:203:9
  17: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:649:30
  18: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:332:26
  19: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:298:17
  20: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:649:30
  21: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:332:26
  22: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:298:17
  23: timely::worker::Wrapper::step::{{closure}}
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/worker.rs:691:57
  24: core::option::Option<T>::map
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:450:29
  25: timely::worker::Wrapper::step
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/worker.rs:691:26
  26: timely::worker::Worker<A>::step_or_park
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/worker.rs:375:38
  27: dataflow::server::Worker<A>::run
             at src/dataflow/src/server.rs:485:13
  28: dataflow::server::serve::{{closure}}
             at src/dataflow/src/server.rs:246:13
  29: timely::execute::execute_from::{{closure}}
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/execute.rs:383:22
  30: timely_communication::initialize::initialize_from::{{closure}}
             at /home/mjibson/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/communication/src/initialize.rs:285:33
  31: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137:18
  32: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:464:17
  33: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:308:9
  34: std::panicking::try::do_call
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:381:40
  35: __rust_try
  36: std::panicking::try
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:345:19
  37: std::panic::catch_unwind
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:382:14
  38: std::thread::Builder::spawn_unchecked::{{closure}}
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:463:30
  39: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /home/mjibson/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  40: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys/unix/thread.rs:87:17
  41: start_thread
  42: __GI___clone
```

**Comment by benesch** [comment]
Ah, yeah, I noticed this over in #5638:

https://github.com/MaterializeInc/materialize/pull/5638/files#diff-7a54f6900c4d8906067a1b8501a9c68145c43ecc97835f27096a3cc24c94b212R35

The problem seems to be that the tail sink assumes the receiver will never drop first, but it can. I believe if the client drops the connection?

The code in #5638 is really hacky, but demonstrates the idea. We could alternatively say the receiver _does_ need to drain entirely, because it's kind of hard for the tail sink to clean itself up when it notices the receiver's gone away.

**Comment by mjibson** [comment]
I just noticed that the close connection branch is never taken because the argument to intn wasn't high enough, so whatever the cause here, it doesn't seem to be due to closed connection.

**Comment by benesch** [comment]
Ah, yeah, with the new declare/fetch stuff I've lost track of how ownership of that tail rx works. It makes sense to me that there are now ways of dropping that tail rx besides closing your connection that is running a tail. Probably committing a transaction will do it, because the cursor gets dropped? Something along those lines, anyway.

**Comment by mjibson** [comment]
The rx is owned by the session portal (it gets turned into a RowBatchStream), so yes, it's dropped at the end of a transaction because that also clears all portals.

**Comment by mjibson** [comment]
The Go program here was made to try to search for more file descriptor leaks (#5543). Fixing the panic here will enable easier automated FD leak detection.

**Comment by benesch** [comment]
@mjibson are you still planning on fixing this? Via draining all portals like we discussed? Apologies if I misremembered and I was supposed to fix it!

**Comment by mjibson** [comment]
It wasn't clear to me who was handling it. I'll do it.

**Comment by mjibson** [comment]
This specific bug was fixed with #5638.
